src/lib.rs

```diff
@@ -6,26 +6,40 @@ use crate::model::{ClusterState, NodeState, ScuttleButtMessage};
 
 // https://www.cs.cornell.edu/home/rvr/papers/flowgossip.pdf
 
+/// HashMap key for the heartbeat node value.
+pub(crate) const HEARTBEAT_KEY: &str = "heartbeat";
+
 pub struct ScuttleButt {
     max_transmitted_key_values: usize, // mtu in the paper
     self_node_id: String,
     cluster_state_map: ClusterState,
+    heartbeat: u64,
 }
 
 impl ScuttleButt {
     pub fn with_node_id(self_node_id: String) -> Self {
-        ScuttleButt {
+        let mut scuttlebutt = ScuttleButt {
             max_transmitted_key_values: 10,
+            heartbeat: 0,
             self_node_id,
             cluster_state_map: ClusterState::default(),
-        }
+        };
+
+        // Immediately mark node as alive to ensure it responds to SYNs.
+        scuttlebutt.self_node_state().set(HEARTBEAT_KEY, 0);
+
+        scuttlebutt
     }
 
     pub fn set_max_num_key_values(&mut self, max_transmitted_key_values: usize) {
         self.max_transmitted_key_values = max_transmitted_key_values;
     }
 
-    pub fn create_syn_message(&self) -> ScuttleButtMessage {
+    pub fn create_syn_message(&mut self) -> ScuttleButtMessage {
+        self.heartbeat += 1;
+        let heartbeat = self.heartbeat;
+        self.self_node_state().set(HEARTBEAT_KEY, heartbeat);
```

Comment:
This is where the heartbeat lives now. I think having it on the `create_syn` makes sense to automagically do it for the consumer regardless of using a server or not.

Thought about adding it to the syn_ack/ack replies too but I feel like a heartbeat shouldn't be incremented when the node isn't capable of creating syns itself anymore.

Reply:
No we want the heartbeat to change on replies too.

(random thought: we could make it into a Lamport clock actually?)

Reply:
> We want the heartbeat to change on replies too.

Sure, that's an easy change.

> (random thought: we could make it into a Lamport clock actually?)

I'm curious what you'd use a Lamport clock for here. IIRC the K/V storage is already versioned and the heartbeat is only ever produced by one node and never consumed by another. So the thing that matters for a heartbeat is the duration since the last time a node was confirmed to be alive, while the actual order of heartbeats doesn't *really* matter.

The only thing I could see is that you'd be able to discard heartbeats that for some reason got severely delayed by the gossiping, which have been produced *after* heartbeats you've already received? But since dead nodes not produce new events, I doubt that would cause them to stick around for **much** longer and they'll eventually be detected as dead anyway.

Reply:
We have no usage for lamport clocks :)

It's just that we need a heartbeat and it can be any value that changes.

- local autoincrement id
- seconds elapsed since startup time
- lamport clock

The lamport clocks might be nice when we debug.
It will make the heartbeat clock comparable, and therefore, make it possible to identify which node is lagging.

On the other hand, seconds elapsed since startup time, is also valuable information for debug.

Reply:
Happy to go with whatever you think fits best here, assuming it doesn't involve a mountain of additional code.

Though there is something to be said about just using the simplest solution possible when it comes to debugging. The easiest bug to fix is the one that was prevented before it was made after all :smile:. Not that any of this code itself is particularly likely to cause troubles.

Reply:
@chrisduerr You are right. Let's stick to seconds elapsed since startup time.